File: app/src/main/java/com/soya/launcher/ui/fragment/AbsLanguageFragment.java
package com.soya.launcher.ui.fragment;

import android.content.res.Resources;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.leanback.widget.ArrayObjectAdapter;
import androidx.leanback.widget.ItemBridgeAdapter;
import androidx.leanback.widget.VerticalGridView;

import com.soya.launcher.R;
import com.soya.launcher.adapter.LanguageAdapter;
import com.soya.launcher.bean.Language;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;

public abstract class AbsLanguageFragment extends AbsFragment implements View.OnClickListener {

    private VerticalGridView mContentGrid;
    private View mNextView;

    @Override
    public int getLayoutId() {
        return R.layout.fragment_set_language;
    }

    @Override
    protected void init(View view, LayoutInflater inflater) {
        super.init(view, inflater);
        mContentGrid = view.findViewById(R.id.content);
        mNextView = view.findViewById(R.id.next);
        setContent(view, inflater);
    }

    @Override
    protected void initBefore(View view, LayoutInflater inflater) {
        super.initBefore(view, inflater);
        mNextView.setOnClickListener(this);
    }

    private void setContent(View view, LayoutInflater inflater){
        List<Language> list = new ArrayList<>();
        LanguageAdapter adapter = new LanguageAdapter(getActivity(), inflater, list);
        ArrayObjectAdapter arrayObjectAdapter = new ArrayObjectAdapter(adapter);
        ItemBridgeAdapter itemBridgeAdapter = new ItemBridgeAdapter(arrayObjectAdapter);
        mContentGrid.setAdapter(itemBridgeAdapter);
        mContentGrid.setColumnWidth(ViewGroup.LayoutParams.WRAP_CONTENT);

        String[] locales = Resources.getSystem().getAssets().getLocales();
        Set<String> strings = new HashSet<>();
        for (int i = 0; i < locales.length; i++) {
            Locale locale = Locale.forLanguageTag(locales[i]);
            String language = locale.getLanguage();
            String name = locale.getDisplayName(locale);
            if (!strings.contains(language) || language.equals("en") || language.equals("zh")) list.add(new Language(name, locale.getDisplayLanguage(), locale));
            strings.add(language);
        }


        /*List<MyLocaleInfo> localeInfos = SystemUtils.getSystemLocaleInfos(getActivity(), false);
        for (int i = 0; i < localeInfos.size(); i++){
            Locale locale = localeInfos.get(i).getLocale();
            String name = locale.getDisplayName(locale);
            list.add(new Language(name, locale.getDisplayLanguage(), locale));
        }*/

        Collections.sort(list, new Comparator<Language>() {
            @Override
            public int compare(Language o1, Language o2) {
                return o1.getName().compareTo(o2.getName());
            }
        });

        int selectLauncher = 0;
        int selectCountry = 0;
        Locale locale = Locale.getDefault();
        for (int i = 0; i < list.size(); i++){
            Locale item = list.get(i).getLanguage();
            if (locale.getLanguage().equals(item.getLanguage())){
                selectLauncher = i;
                if (locale.getCountry().equals(item.getCountry())){
                    selectCountry = i;
                }
            }
        }
        int select = selectCountry != 0 ? selectCountry : selectLauncher;
        adapter.setSelect(select);

        adapter.setCallback(newCallback(arrayObjectAdapter));

        arrayObjectAdapter.addAll(0, list);
        mContentGrid.setSelectedPosition(select);
    }

    public LanguageAdapter.Callback newCallback(ArrayObjectAdapter adapter){
        return new LanguageAdapter.Callback() {
            @Override
            public void onClick(Language bean) {
                onSelectLanguage(bean);
                adapter.notifyArrayItemRangeChanged(0, adapter.size());
            }
        };
    }

    @Override
    protected int getWallpaperView() {
        return R.id.wallpaper;
    }

    @Override
    public void onClick(View v) {
        if (v.equals(mNextView)){
            onNext();
        }
    }

    protected void showNext(boolean show){
        mNextView.setVisibility(show ? View.VISIBLE : View.GONE);
    }

    protected void onNext(){}

    protected void onSelectLanguage(Language bean){}
}

File: app/src/main/java/com/soya/launcher/ui/fragment/AbsLanguageFragment.kt
package com.soya.launcher.ui.fragment

import android.content.res.Resources
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.ViewDataBinding
import androidx.leanback.widget.ArrayObjectAdapter
import androidx.leanback.widget.ItemBridgeAdapter
import androidx.leanback.widget.VerticalGridView
import com.drake.brv.utils.addModels
import com.drake.brv.utils.mutable
import com.drake.brv.utils.setup
import com.shudong.lib_base.base.BaseViewModel
import com.shudong.lib_base.ext.appContext
import com.shudong.lib_base.ext.clickNoRepeat
import com.soya.launcher.BaseWallPaperFragment
import com.soya.launcher.R
import com.soya.launcher.bean.Language
import com.soya.launcher.bean.SimpleTimeZone
import com.soya.launcher.databinding.FragmentSetLanguageBinding
import com.soya.launcher.databinding.HolderLanguageBinding
import com.soya.launcher.databinding.HolderTimeZoneBinding
import java.util.Collections
import java.util.Locale

abstract class AbsLanguageFragment<VDB : FragmentSetLanguageBinding, VM : BaseViewModel> :
    BaseWallPaperFragment<VDB, VM>(), View.OnClickListener {


    override fun initView() {
        setContent()
        mBind.content.post {
            mBind.content.requestFocus()
        }

        mBind.next.setOnClickListener(this)
    }


    private var selectIndex = 0
    private fun setContent() {
        mBind.content.setup {
            addType<Language>(R.layout.holder_language)
            onBind {
                val dto = getModel<Language>()
                val binding = getBinding<HolderLanguageBinding>()
                val isSelect = modelPosition == selectIndex
                binding.check.visibility = if (isSelect) View.VISIBLE else View.GONE
                binding.desc.visibility = if (isSelect) View.GONE else View.VISIBLE
                itemView.clickNoRepeat {
                    selectIndex = modelPosition
                    onSelectLanguage(dto)
                    notifyItemRangeChanged(0, mBind.content.mutable.size)
                }
            }
        }
        val list: MutableList<Language> = ArrayList()
        mBind.content.setColumnWidth(ViewGroup.LayoutParams.WRAP_CONTENT)

        val locales = Resources.getSystem().assets.locales
        val strings: MutableSet<String> = HashSet()
        for (i in locales.indices) {
            val locale = Locale.forLanguageTag(locales[i])
            val language = locale.language
            val name = locale.getDisplayName(locale)
            if (!strings.contains(language) || language == "en" || language == "zh") list.add(
                Language(name, locale.displayLanguage, locale)
            )
            strings.add(language)
        }

        list.sortWith { o1, o2 -> o1.name.compareTo(o2.name) }

        var selectLauncher = 0
        var selectCountry = 0
        val locale = Locale.getDefault()
        for (i in list.indices) {
            val item = list[i].language
            if (locale.language == item.language) {
                selectLauncher = i
                if (locale.country == item.country) {
                    selectCountry = i
                }
            }
        }
        val select = if (selectCountry != 0) selectCountry else selectLauncher
        selectIndex = select
        mBind.content.addModels(list)
        mBind.content.selectedPosition = select
    }


    override fun onClick(v: View) {
        if (v == mBind.next) {
            onNext()
        }
    }

    protected fun showNext(show: Boolean) {
        mBind.next.visibility = if (show) View.VISIBLE else View.GONE
    }

    protected open fun onNext() {}

    protected open fun onSelectLanguage(bean: Language) {}
}
